src/HTSPTypes.cpp

```diff
@@ -437,3 +441,337 @@ void AutoRecording::SetFulltext(uint32_t fulltext)
+  if (m)
+    htsmsg_destroy(m);
+}
+
+void Subscription::SendWeight(uint32_t weight)
+{
+  CLockObject lock(m_conn.Mutex());
+  if (!IsActive() || GetWeight() == weight)
+    return;
+
+  htsmsg_t *m;
+  m_weight = weight;
+
+  /* Build message */
+  m = htsmsg_create_map();
+  htsmsg_add_u32(m, "subscriptionId", m_subscriptionId);
+  htsmsg_add_s32(m, "weight",         m_weight);
+  tvhdebug("demux send weight %u",    m_weight);
+
+  /* Send and Wait */
+  m = m_conn.SendAndWait("subscriptionChangeWeight", m);
+  if (m)
+    htsmsg_destroy(m);
+}
+
+void Subscription::ParseSubscriptionSpeed (htsmsg_t *m)
+{
+  uint32_t u32;
+  if (!htsmsg_get_u32(m, "speed", &u32))
+    tvhtrace("recv speed %d", u32);
+}
+
+void Subscription::ParseSubscriptionStatus (htsmsg_t *m)
+{
+  /* Not for preTuning and postTuning subscriptions, tag them as unknown */
+  if (GetWeight() == SUBSCRIPTION_WEIGHT_PRETUNING ||
+      GetWeight() == SUBSCRIPTION_WEIGHT_POSTTUNING)
+  {
+    SetState(SUBSCRIPTION_UNKNOWN);
+    return;
+  }
+
+  const char *status, *error;
+
+  /* 'status is used by older tvheadend servers */
+  status = htsmsg_get_str(m, "status");
+
+  if (status != NULL)
+    tvhinfo("Bad subscription status: %s", status);
+
+  /* 'subscriptionError' is the new alternative for 'status' since htsp v20 */
+  error  = htsmsg_get_str(m, "subscriptionError");
+
+  if (m_conn.GetProtocol() >= 20)
+  {
+    /* This field is absent when everything is fine */
+    if (error != NULL)
+    {
+      if (!strcmp("badSignal", error))
+        SetState(SUBSCRIPTION_NOSIGNAL);
+      else if (!strcmp("scrambled", error))
+        SetState(SUBSCRIPTION_SCRAMBLED);
+      else if (!strcmp("userLimit", error))
+        SetState(SUBSCRIPTION_USERLIMIT);
+      else if (!strcmp("noFreeAdapter", error))
+      {
+        if (GetState() != SUBSCRIPTION_NOFREEADAPTER &&
+            GetState() != SUBSCRIPTION_FORCERUNNING_FAILED)
+        {
+          /* No free adapter, AKA subscription conflict, show OSD dialog */
+          std::thread job(&Subscription::ShowConflictDialog, this);
+          job.detach();
+
+          SetState(SUBSCRIPTION_NOFREEADAPTER);
+        }
+      }
+      else
+        SetState(SUBSCRIPTION_UNKNOWN);
+
+      /* Show an OSD message */
+      ShowStateNotification();
+    }
+    else
+      SetState(SUBSCRIPTION_RUNNING);
+  }
+  else
+  {
+    /* This field is absent when everything is fine */
+    if (status != NULL)
+    {
+      XBMC->QueueNotification(ADDON::QUEUE_INFO, status);
+      SetState(SUBSCRIPTION_UNKNOWN);
+    }
+    else
+      SetState(SUBSCRIPTION_RUNNING);
+  }
+}
+
+/* Threadsafe getter for m_state */
+eSubsriptionState Subscription::GetState(void)
+{
+  m_mutex.Lock();
+  eSubsriptionState state = m_state;
+  m_mutex.Unlock();
+  return state;
+}
+
+/* Threadsafe setter for m_state */
+void Subscription::SetState(eSubsriptionState state)
+{
+  m_mutex.Lock();
+  m_state = state;
+  m_mutex.Unlock();
+}
+
```

Comment:
You could take a look at https://github.com/kodi-pvr/pvr.hts/blob/master/src/AsyncState.cpp and https://github.com/kodi-pvr/pvr.vbox/blob/master/src/vbox/StartupStateHandler.h. Your basic idea is correct, you need to use a mutex lock whenever you access the `m_state` variable. `const' functions are not thread-safe by definition since another thread could come in and overwrite the value of the variable while you're reading it. Since locking a mutex is a non-const operation it is common to mark the mutex variable as mutable so it can be used from `const` methods (see e.g. https://github.com/kodi-pvr/pvr.vbox/blob/master/src/vbox/StartupStateHandler.h#L123).